First cut of agent diagnose CLI command extension

iglendd opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.
-->
### What does this PR do?
<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->
Extended agent diagnose CLI command internal implementation to allow a simple registration of a new diagnose "suite" which will return one or more diagnoses by any part of agent's code.

#### CLI changes
From the CLI standpoint most of the changes are covered by the new **agent diagnose all** sub-command. Other, previous sub-commands, left AS IS for now for compatibility reasons. But further changes are possible, e.g. removal of all sub-commands and using only command-line options to fine-tune diagnose depth, breadth and output format. Suggestions are welcome.

New command line options/flags for the **agent diagnose all** sub-command
```
  -e, --exclude strings   diagnose suites not to run as a list of regular expressions
  -i, --include strings   diagnose suites to run as a list of regular expressions
  -l, --list              list diagnose suites
  -u, --run-as-user       run as current user
  -v, --verbose           verbose output, includes passed diagnoses, and diagnoses description
```
#### Overall changes
* Basic infrastructure and interface for extending agent diagnose command within Agent proper.
* Light refactoring of existing diagnose CLI sub-commands.
* Few additional concreate diagnose tests, including DBM and NDM connectivity diagnostics
* Python check interface extended to include new `GetDiagnoses()` method to facilitate diagnose functionality in the python check code.

#### Architectural changes
Reasoning, architectural changes and additional details can be found in th [RFC: Extending Agent Diagnose](https://docs.google.com/document/d/181y_yh4IwvT6xvEGDvsOTkwoXkO7WwhkV4182OC2VJE/edit#heading=h.i32aowyibiwe)
![image](https://user-images.githubusercontent.com/98782554/235527907-dfb7916b-2881-4582-974f-3ba072511c5d.png)

### Motivation
<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->
To simplify and promote further agent diagnose command extensions via concrete diagnosis to help to self-diagnose and self-troubleshooting agent problems.

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->
Changed files count is relatively large, **51 files**, but most of the changes are to propagate interface/function changes and most of the files have 1 or 2 lines changes. Only **14 files** have more substantial changes.

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes
Run permutations of agent diagnose command. To see options use
`agent diagnose -h` help to see these options. New command is `agent diagnose all`. Use  `agent diagnose all -h` to see its options (not all options are supported at this time). Other  previous diagnose sub-command should continue to operate as before.

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### README.md
#### ```agent diagnose``` CLI command

```agent diagnose``` command is one of the Agent CLI commands ([agent other commands](https://docs.datadoghq.com/agent/guide/agent-commands/?tab=agentv6v7#other-commands)). It run  internal self-diagnostic tests and outputs their results and if problems are found report them (and in some cases remediation steps). 

##### ```diagnose``` command line options
List sub-commands:
```
agent diagnose -h
Validate Agent installation, configuration and environment

Usage:
  agent.exe diagnose [flags]
  agent.exe diagnose [command]

Available Commands:
  all                   Validate Agent installation, configuration and environment
  datadog-connectivity  Check connectivity between your system and Datadog endpoints
  metadata-availability Check availability of cloud provider and container metadata endpoints
  show-metadata         Print metadata payloads sent by the agent
```
##### ```diagnose all``` sub-command
###### Options
List all sub-command options:
```
agent.exe diagnose all -h
Validate Agent installation, configuration and environment

Usage:
  agent.exe diagnose all [flags]

Flags:
  -e, --exclude strings   diagnose suites not to run as a list of regular expressions
  -h, --help              help for all
  -i, --include strings   diagnose suites to run as a list of regular expressions
  -l, --list              list diagnose suites
  -u, --run-as-user       run as current user
  -v, --verbose           verbose output, includes passed diagnoses, and diagnoses description
```

###### ```include``` and ```exclude``` options
agent diagnose --include and or --exclude options allow to filter to specific diagnose suites

###### ```list``` option
List names of all registered diagnose suites. Output also will be filtered if include and or exclude options are specified.

###### ```verbose``` option
Normally a successful diagnosis is printed as a single dot character. If verbose option is specified successful diagnosis is printed fully. With verbose option diagnosis description is also printed.

###### ```run-as-user``` option
Normally internal diagnose functions will run in the context of agent and other services. It can be overridden via –run-as-user options and if specified diagnose functions will be executed in context of the agent diagnose CLI process if possible.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

iglendd committed:
First cut of base agent diagnose extension

* Basic infrastructure
* Interfaces
* Some refactoring of exitinging diagnose CLI commands
* Few additional diagnose tests, including DBM and NDM
* Interface for diagnose in Python checks

maycmlee commented on releasenotes/notes/agent-diagnose-ext-first-cut-0b73b9442cf173e3.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+features:
+  - |
+    Extended "agent diagnose" CLI command to allow easy addition of new
+    diagnostics from dispersed agent code.
```
```suggestion
    Extended the "agent diagnose" CLI command to allow the easy addition of new
    diagnostics from dispersed Agent code.
```

iglendd replied:
@maycmlee thank you for the suggestion I have accepted it largely but added few more clarifications. It is now
```
    Extended the "agent diagnose" CLI command to allow the easy addition of new
    diagnostics for diverse and dispersed Agent code.
```
which better reflect what I was tried to explain. Please let me know if it is still sufficient.

maycmlee reviewed (approved):
A couple of suggestions, but approving.

iglendd committed:
Added two clarifying comments, including from code review suggestion.

maycmlee reviewed (approved):
LGTM!

just-chillin reviewed (approved):
Exciting to see these things unfold! :shipit:

katiehockman commented on pkg/diagnose/connectivity/trace_request.go:
```diff
@@ -24,10 +24,74 @@ import (
+		}
+	}
+
+	diagnoses := make([]diagnosis.Diagnosis, 0)
```
```suggestion
	var diagnoses []diagnosis.Diagnosis
```
any reason not to just define this as a var if it has a zero length?

iglendd replied:
Only my ignorance. I am go newbie. I thought that slice need to be explicitly initialized otherwise it would be nil and panic if appended to. Tested your suggestion in Go playground and works well. Thank you.

iglendd replied:
I have pushed the fix BTW

katiehockman commented on pkg/diagnose/connectivity/trace_request.go:
```diff
@@ -24,10 +24,74 @@ import (
+}
+
+func diagnose(diagCfg diagnosis.DiagnoseConfig) []diagnosis.Diagnosis {
+
```
```suggestion
```
nit: there's a lot of new lines in this function that I'm not sure are necessary to break apart the functionality of the code. I'd suggest reducing them a bit (e.g. 53, 55, 57 can also probably go)

iglendd replied:
I agree I went bit too far with spaces in this function. Cannot recall what I was thinking. Probably did not clean properly debugging code. Thank you. I have removed few of them.

katiehockman commented on pkg/diagnose/connectivity/trace_request.go:
```diff
@@ -24,10 +24,74 @@ import (
+
+				domain, _ := domainResolver.Resolve(endpointInfo.Endpoint)
+
+				ctx := context.Background()
```
does every request to an endpoint need a new context? Or can we just define a context at the beginning of the function and re-use it throughout?

iglendd replied:
I fee that new context needed at least on the face of it since it is formed from the inner loop object. Perhaps better informed Agent developer like @sgnn7 or @truthbk  can suggest if this is possible.

truthbk replied:
Given that it doesn't seem like there'll be concurrency across requests I think using a single `ctx` might be OK, actually. We don't do much with the context even in the face of a timeout or whatever either.

iglendd replied:
I have looked at the wrong variable endpointInfo for some reason. If course ctx should be moved out. Done (in the next commit).

katiehockman replied:
LGTM 👍

katiehockman commented on pkg/diagnose/connectivity/trace_request.go:
```diff
@@ -24,10 +24,74 @@ import (
+
+				// Check if there is a response and if it's valid
+				report, reportErr := verifyEndpointResponse(statusCode, responseBody, err)
+				name := "Connectivity to " + logURL
+				if reportErr == nil {
+					diagnoses = append(diagnoses, diagnosis.Diagnosis{
+						Result:    diagnosis.DiagnosisSuccess,
+						Name:      name,
+						Diagnosis: fmt.Sprintf("Connectivity to `%s` is Ok\n%s", logURL, report),
+					})
+				} else {
+					diagnoses = append(diagnoses, diagnosis.Diagnosis{
+						Result:      diagnosis.DiagnosisFail,
+						Name:        name,
+						Diagnosis:   fmt.Sprintf("Connection to `%s` is falied\n%s", logURL, report),
+						Remediation: "Please validate Agent configuration and firewall to access " + logURL,
+						RawError:    err,
+					})
+				}
```
```suggestion
				d := diagnosis.Diagnosis{
					Name: "Connectivity to " + logURL
				}
				if reportErr == nil {
					d.Result = diagnosis.DiagnosisSuccess
					d.Diagnosis = fmt.Sprintf("Connectivity to `%s` is Ok\n%s", logURL, report)
				} else {
					d.Result = diagnosis.DiagnosisFail
					d.Diagnosis = fmt.Sprintf("Connection to `%s` is falied\n%s", logURL, report)
					d.Remediation = "Please validate Agent configuration and firewall to access " + logURL
					d.RawError = err
				}
				diagnoses = append(diagnoses, d)
```
optional style change

iglendd replied:
I like it. I did not particular like my construction. I am missing ternary expressions. Thank you.

iglendd committed:
Core review related changes

hush-hush commented on rtloader/three/three.cpp:
```diff
@@ -520,6 +520,7 @@ char **Three::getCheckWarnings(RtLoaderPyObject *check)
         if (warn == NULL) {
             setError("there was an error browsing 'warnings' list: " + _fetchPythonError());
 
+            // Is this a bug? ---> if warnings[jdx] != NULL and jdx > idx (uninitialized memory)
```
Seems like it indeed ! We might want to fix this in another PR. I'll create a card in our backlog.

vickenty replied:
Could you please remove this comment? https://github.com/DataDog/datadog-agent/pull/17490 will fix the bug, and this line will cause merge conflicts.

hush-hush commented on rtloader/two/two.cpp:
```diff
@@ -529,6 +529,137 @@ char **Two::getCheckWarnings(RtLoaderPyObject *check)
     return warnings;
 }
 
+diagnoses_t *Two::getCheckDiagnoses(RtLoaderPyObject *check)
```
This looks like the same implementation between `two` and `three`. We might want to move logic to a helper function ?

iglendd replied:
I can do it and I was not particular thrilled about it but it loos like bunch of two and three methods are mostly copy/paste. I do not know if it is a pattern I should observe and if there are good reasons for that and I was trying to follow existing pattern. Correct me if I am wrong here.

hush-hush replied:
The issue with other method was that there were just enough difference between two and three. In the end making shared code between the two was more trouble that it was worth it. But in this case the code seems exactly the same (I guess Python list API didn't changed from 2 to 3). I think it's worth making an helper function.

iglendd replied:
Ok, I will move these virtual methods in the base class.

iglendd replied:
Looks it is a bit involved for comfort in terms of include dependencies. Will move common code into common/diagnose.c/h files

iglendd replied:
That was painful. I have tried a few things. I could not move to the base because it appears to be Python version neutral and I could not have a general rtloader general CPP code. I have ended up creating two lower level size calculation and serialization functions. I did not eliminate duplicate code completely by adding a good chunk of it to the util code. Let me know if it is good that I can resolve the comment.

hush-hush commented on pkg/diagnose/check.go:
```diff
@@ -0,0 +1,120 @@
+	diagnosis.Register("check-datadog", diagnose)
+}
+
+// Diagnose configured checks. Run them twice to facilitate rate-based metrics
+// which otherwise will fail. In the future perhaps it can be avoided for
+// non-rate checks. Currently diagnose is implemented to run in the CLI process,
+// in the next version will connect to the running agent service to get diagnoses
+// for scheduled checks.
```
Do we need to run the check twice for diagnose ?

iglendd replied:
Old comment. I need to fix it. It was original implementation and @alopezz already mention that and I have addressed. Will fix the comment. Thank you.

hush-hush commented on pkg/diagnose/check.go:
```diff
@@ -0,0 +1,120 @@
+
+			defaultDiagnosisName := fmt.Sprintf("Check: %s [%s]", checkName, string(instance.ID()))
+
+			// Run check
+			runErr := instance.Run()
```
Do we need to run checks to call their diagnose method ?

iglendd replied:
Very observant. No, I do not. And I am not planning to in future when I will call checks in context of agent service. I do now because I have not implemented that and also to allow @alopezz to test his python check with this release. But it is not easy to spot that nuance for a code reviewer. Thank you.

hush-hush reviewed (commented):
Look great overall. I added some comments here and there but nothing major to say.
The PR being huge, we might want another look, especially from @DataDog/agent-metrics-logs on the C part.

iglendd commented on cmd/agent/subcommands/diagnose/README.md:
````diff
@@ -0,0 +1,50 @@
+### ```verbose``` option
+Normally a successful diagnosis is printed as a single dot character. If verbose option is specified successful diagnosis is printed fully. With verbose option diagnosis description is also printed.
+
+### ```run-as-user``` option
+Normally internal diagnose functions will run in the context of agent and other services. It can be overridden via –run-as-user options and if specified diagnose functions will be executed in context of the agent diagnose CLI process if possible.
````
@hush-hush Thank you. It looks like it is an easy to misinterpret comment and terminology. What I was trying to communicate here is that normally (by default) most of the agent diagnose code runs inside (or within context) of agent(.exe) service. But I also want to give a user a choice to run the very same diagnose code under privilege of a user which started this CLI agent diagnose command meaning to not invoke remote (from CLI command perspective) diagnose code but run it directly (just like agent check does for checks). For that I introduced –run-as-user but you are right the term is too similar named as the one which is used to run a process under privileges of an arbitrary user which was not my intention.

What other option name should I choose for that.--run-as-current-user maybe?

iglendd replied:
Perfect. Did not know that. Absolutely agree. Changing it now 🙏

iglendd commented on cmd/agent/subcommands/diagnose/README.md:
```diff
@@ -0,0 +1,50 @@
+  -e, --exclude strings   diagnose suites not to run as a list of regular expressions
+  -h, --help              help for all
+  -i, --include strings   diagnose suites to run as a list of regular expressions
+  -l, --list              list diagnose suites
```
This is where is probably made a wrong decision from the very beginning. I have decided, afraid to change/break existing CLI, to keep "new" diagnose CLI interface separate from what we already have. I did not know exactly how "far" I will or should I go and certainly did not want to jeopardize of what is already there. I feel right now, and I have implied that in the PR comment, that I should not have "all" as segregated CLI command and instead integrate existing CLI commands into existing framework at implementation and at CLI level. In that case your suggestion would be even relevant in a sense that that would the only possible way to deal with that. May be I should take another swing into that (in a separate PR). Not sure if I explain it well though.

iglendd replied:
@hush-hush and @Kaderinho and perhaps @sgnn7 any comment of the overall diagnose all subcommand. Should we have a meeting to discuss what I have in PR vs "ideal"?

iglendd replied:
Three of us had a meeting and have decided to "harmonize" agent diagnose CLI interface and output in the separate PR

iglendd commented on pkg/diagnose/diagnosis/loader.go:
```diff
@@ -5,21 +5,102 @@
+// Diagnose interface function
+type Diagnose func(DiagnoseConfig) []Diagnosis
+
+// Global list of registered Diagnose functions
+var DiagnoseCatalog = make([]DiagnoseSuite, 0)
+
+// Diagnose suite information
+type DiagnoseSuite struct {
+	SuitName string
+	Diagnose Diagnose
+}
+
+// Diagnose configuration
+type DiagnoseConfig struct {
```
I am not sure which lines are you referring exactly. I did not change it but I did find bit odd existence of two packages inside a single directory `diagnose` for most of the `datadog-agent\pkg\diagnose`  directory and subdirectory and `diagnosis` for a single file `datadog-agent\pkg\diagnose\diagnosis\loader.go`. Note a difference between diagnose and diagnosis (verb vs noun). Again, this was not my doing I thought that was probably done to break some circular dependencies and simplify partitioning on the interface and simple data. I could be wrong. If I could remove some diagnosis prefix from within diagnose packages I would be glad but it does not seem to be possible. Unless I misunderstand your suggestion.

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+}
+
+// Overall running statistics
+type diangosisCounters struct {
+	totalCnt   int
+	successCnt int
+	failCnt    int
+	warningCnt int
+	errorCnt   int
+	skippedCnt int
+}
```
Good idea.

iglendd replied:
I think I started with errorCnt since error is reserved word and copied it everywhere after.

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+		color.NoColor = true
+	}
+
+	// Sort Diagnose by their suite names
+	sortedSuites := make([]diagnosis.DiagnoseSuite, len(diagnosis.DiagnoseCatalog))
+	copy(sortedSuites, diagnosis.DiagnoseCatalog)
+	sort.Slice(sortedSuites, func(i, j int) bool {
+		return sortedSuites[i].SuitName < sortedSuites[j].SuitName
+	})
```
I think I can. Good idea.

iglendd replied:
I probably will hold off for this for now for two reason. a) I do not want to resort after each registration and I do not have a terminal signal b) I am not sure if I want to sort always by this.

I will make a single sorting function though. Thank you.

iglendd commented on rtloader/include/rtloader_types.h:
```diff
@@ -72,6 +72,40 @@ typedef struct event_s {
+// Self-contained array of diagnosis. All pointers point to portion of memory contained
+// completely within the diagnoses_t buffer. size_t or pointers are needed to avoid Go vs. CGO misalignments
+typedef struct diagnoses_s {
+    size_t byteCout;
```
You are right. I was planning to add it but forgot. I think it is useful still.

iglendd replied:
yes https://github.com/DataDog/datadog-agent/blob/9db2783488d8fff246a9ec742c6b087472f9c7d8/rtloader/three/three.cpp#L595

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+		// Run partiocular diagnose
+		diagnoses := ds.Diagnose(diagCfg)
+		if diagnoses == nil {
+			// No diagnoses are reported, move on to next Diagnose
```
I think I can add it in verbose mode. Good suggestion.

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+	skippedCnt int
+}
+
+// Output summary
+func outputSummary(w io.Writer, c diangosisCounters) {
+	fmt.Fprintf(w, "-------------------------\n  Total:%d", c.totalCnt)
+	if c.successCnt > 0 {
+		fmt.Fprintf(w, ", Success:%d", c.successCnt)
+	}
+	if c.failCnt > 0 {
+		fmt.Fprintf(w, ", Fail:%d", c.failCnt)
+	}
+	if c.warningCnt > 0 {
+		fmt.Fprintf(w, ", Warning:%d", c.warningCnt)
+	}
+	if c.errorCnt > 0 {
+		fmt.Fprintf(w, ", Error:%d", c.errorCnt)
+	}
+	fmt.Fprint(w, "\n")
+}
+
+func incrementCounters(r diagnosis.DiagnosisResult, c *diangosisCounters) {
```
I do not have strong preference either way. The only arguably small (conceptual) flaw in you suggestion, IMHO, that dummy struct know how display itself into output. But if you feel that overall it is better approach I do not have problem of changing it.

iglendd replied:
With a fresher mind I have decided to change as you have suggested. It is in single file and my concern is not really important for that local scope.

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+	}
+}
+
+func getDiagnosisResultForOutput(r diagnosis.DiagnosisResult) string {
+	var result string
+	if r == diagnosis.DiagnosisSuccess {
+		result = color.GreenString("PASS")
+	} else if r == diagnosis.DiagnosisFail {
+		result = color.RedString("FAIL")
+	} else if r == diagnosis.DiagnosisWarning {
+		result = color.YellowString("WARNING")
+	} else { //if d.Result == diagnosis.DiagnosisUnexpectedError
+		result = color.HiRedString("UNEXPECTED ERROR")
+	}
+
+	return result
+}
```
I have the same similar and small concern of coupling behavior and structure. Not particular strong and I do not mind changing it since it is a bit of a grey area in that regard I think

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+	}
+}
+
+func outputDot(w io.Writer, lastDot *bool) {
+	fmt.Fprint(w, ".")
+	*lastDot = true
+}
```
Will try to rethink it. I also not particular in love with necessity of this (never mind - I do hate that part). The problem I was solving was non-verbose output where I print a dot for each successful diagnosis but I need to break from the line of dots on a failed diagnosis.

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+				continue
+			}
+
+			outputSuiteIfNeeded(w, ds.SuitName, &suiteAlreadyReported)
```
It is related to the dot problem. But perhaps your idea of always showing suite is sufficient. I was trying to minimize real-estate for successful diagnosis, including diangose suites. Imaging if we have few dozens of them of more in future. Half of the screen of more will be occupied by names of the suites. It is arguable position of course but that how I feel about it.

iglendd replied:
Actually that exactly what I do. In verbose I do print suite name. I have to sleep on it if I can improve code implementing desired output is less hacky and bit more elegant.

iglendd replied:
Will try to address this logic in the subsequent PR

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+// Output summary
+func outputSummary(w io.Writer, c diangosisCounters) {
+	fmt.Fprintf(w, "-------------------------\n  Total:%d", c.totalCnt)
+	if c.successCnt > 0 {
+		fmt.Fprintf(w, ", Success:%d", c.successCnt)
+	}
+	if c.failCnt > 0 {
+		fmt.Fprintf(w, ", Fail:%d", c.failCnt)
+	}
+	if c.warningCnt > 0 {
+		fmt.Fprintf(w, ", Warning:%d", c.warningCnt)
+	}
+	if c.errorCnt > 0 {
+		fmt.Fprintf(w, ", Error:%d", c.errorCnt)
+	}
+	fmt.Fprint(w, "\n")
```
I think there is no big harm in it but it is similar to other reason - I do not want to occupy visual realestate and mental load. Consider the case if we need to add few more specific error (e.g. not enough privileges). And do we want to show 0 warnings or zero internal errors when there are none. I think it is a small price to pay with few if statements. I could be wrong though.

iglendd replied:
@hush-hush I have a hard time imagining 
Success: x, Error: 0, Warning: 0, Internal Error: 0
for most of the cases. If it is only to eliminate few if I think it is a waste.

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+	if c.successCnt > 0 {
+		fmt.Fprintf(w, ", Success:%d", c.successCnt)
+	}
+	if c.failCnt > 0 {
```
I need to rename error to better reflect DiagnosisUnexpectedError which I did for result. I am thinking to differentiate between diagnose code detected problem vs diagnose code failed unexpectedly and miserably.

hush-hush replied:
What about `issue` for detected problems and `error` for core  failures ?

iglendd replied:
From high level conceptual point of view we need to make it clear to user and devs who are going to implement that how one can explain the difference and purpose in context of a diagnosis result. I am suggesting `fail` and `internalError`. To me `issue` and `error` does not clarify it well but who am I regarding English word connations. We should ask few more people. @sgnn7 and @truthbk can you comment on this?

iglendd commented on cmd/agent/subcommands/diagnose/command.go:
```diff
@@ -47,12 +67,49 @@ func Commands(globalParams *command.GlobalParams) []*cobra.Command {
+	diagnoseAllCommand.PersistentFlags().BoolVarP(&cliParams.runAsUser, "run-as-user", "u", false, "run as current user")
+
+	// List of regular expressions to include and or exclude names of diagnose suites
+	diagnoseAllCommand.PersistentFlags().StringSliceVarP(&cliParams.include, "include", "i", []string{}, "diagnose suites to run as a list of regular expressions")
+	diagnoseAllCommand.PersistentFlags().StringSliceVarP(&cliParams.exclude, "exclude", "e", []string{}, "diagnose suites not to run as a list of regular expressions")
```
Where? In the comment of in the help string? I have problem with shortness of the help string. I think I also adopted the help string from other location.

iglendd replied:
Wait a second, why mutually exclusive. I believe I have seen in some check when both are used although one need to specify eval order. For example consider following input list

- aaa111zzz
- aaa111yyy
- aaa111xxx
- bbb111zzz

for this case one should be able to narrow it to --include aaa --exclude zzz which should leave only
- aaa111yyy
- aaa111xxx

Kaderinho replied:
I read this file before the actual implementation and thought that we could not use both options at the same time (and forgot to remove the comment after) so please ignore it.

hush-hush replied:
My bad, I must have miss-read the code. I think I miss the first if case here https://github.com/DataDog/datadog-agent/pull/16719/files#diff-7490fc6d3c928262876d9ad313837f636bca99ba772e5c448ad4e9619dddd1e0R162.

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+
+// Enumerate registered Diagnose suites and get their diagnoses
+// for structural output
+func RunAll(diagCfg diagnosis.DiagnoseConfig) []diagnosis.Diagnoses {
```
It will be used when I will add a structured remote call to agent service (but technically you are correct for this PR).

iglendd replied:
... and I completely forgot. I actually added it to use also in the unit tests

iglendd commented on pkg/epforwarder/epforwarder.go:
```diff
@@ -190,6 +198,53 @@ func (s *defaultEventPlatformForwarder) SendEventPlatformEvent(e *message.Messag
+		endpoints, err := config.BuildHTTPEndpointsWithConfig(configKeys, desc.hostnameEndpointPrefix, desc.intakeTrackType, config.DefaultIntakeProtocol, config.DefaultIntakeOrigin)
+		if err != nil {
+			diagnoses = append(diagnoses, diagnosis.Diagnosis{
+				Result:      diagnosis.DiagnosisSuccess,
```
Yep, good find. Thank you.

iglendd commented on releasenotes/notes/agent-diagnose-ext-first-cut-0b73b9442cf173e3.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+features:
+  - |
+    Extended the "agent diagnose" CLI command to allow the easy addition of new
+    diagnostics for diverse and dispersed Agent code.
```
I am not sure in this PR there were only a small number and mostly added was a mechanism. Are our release notes more for end users instead of people with dev inclinations.

iglendd replied:
Sure. I can make the change but can we discuss if diagnose all is suitable change by itself or in the longer run we want to change it?

iglendd replied:
Will make the change in the subsequent PR because I will change command CLI

iglendd commented on pkg/collector/python/check.go:
```diff
@@ -330,6 +332,91 @@ func (c *PythonCheck) ID() check.ID {
+	// Lock the GIL and release it at the end of the run (will crash otherwise)
+	gstate, err := newStickyLock()
+	if err != nil {
+		return nil, nil
```
Yes, I do not know what I was thinking. I think I was experimenting with this only to start working and forgot to address it. Thank you.

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -8,30 +8,29 @@
  * \return A standard C string (NULL terminated character pointer)
- *  The returned pointer is allocated from the heap and must be
- * deallocated (free()ed) by the caller
+ *  The returned pointer is embedded witin PyObject. When string pointer
```
Good find, thank you.

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -8,30 +8,29 @@
 /**
  * returns a C (NULL terminated UTF-8) string from a python string.
  *
  * \param object  A Python string to be converted to C-string.
```
will do

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -44,10 +43,12 @@ char *as_string(PyObject *object)
-    PyObject *temp_bytes = NULL;
 
+    temp_bytes = object;
+    retval = tmp;
```
Not sure I understand. It is but it is only used there not in "three" section. Probably the function is not the most elegant but I did kind of cobble up together from other pieces of code I have found.

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -65,14 +66,161 @@ char *as_string(PyObject *object)
+    return retval;
+}
+
+char *attr_as_embedded_string(PyObject *object, const char *attributeName, PyObject **stringObject)
```
You are right. It was used before (when I had serialized all objects into a single buffer) but not anymore. I will remove it.

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -8,30 +8,29 @@
+ *  is not needed corresponding PyObject reference needs to be decremented
  */
-char *as_string(PyObject *object)
+char *as_embedded_string(PyObject *object, PyObject **stringObject)
```
Yes, this is a key piece in the algorithm of having copying Python strings into single destination buffer without their allocation and without "harder" code work on Python buffer string direct access. This is a "trick" to avoid intermedeate allocation while keeping most of the exiting string manipulation code. This is a way to "pin" the python string object and have a safety of direct access to its underlying char* to copy it at will. Only then I deref it. Just look where this function is called from to see it. I consider myself lucky to find this relatively simple approach without major code surgery to achieve desired effect - no intermediate memory allocation and still relatively simple to follow code (IMHO).

hush-hush replied:
OK, I see !
Could you add a comment to the function explaining this ?

iglendd replied:
@hush-hush I have added another comment. @truthbk new function allows for outside of stringutils to get length and copy string to other buffer if needed without need to track Python object - it is all done inside stringutils. Old function still exists and used by lot of existing code. Yes, new code is bit riskier but I feel bad keep potentially allocating dozens small intermediate buffer and explicitly freeing across CGO boundary.

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -44,10 +43,12 @@ char *as_string(PyObject *object)
+    temp_bytes = object;
+    retval = tmp;
+    Py_INCREF(temp_bytes);
+    *stringObject = temp_bytes;
```
I copied these parts from existing places nearly AS IS. I am not that expert on CPyton nuances.

iglendd replied:
I did notice small difference between 2 and 3 and tried to replicate it instead of rewriting it without understanding full implications in CPython internals.

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -65,14 +66,161 @@ char *as_string(PyObject *object)
+            size = strlen(tmp) + 1;
+            Py_XDECREF(temp_bytes);
+        }
+    } else if (py_attr != NULL && !PyUnicode_Check(py_attr)) {
+        PyErr_Clear();
+    } else {
+        PyErr_Clear();
+    }
```
Yes, it is obvious now (but not when I have written it :)) thank you. Will fix it.

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -65,14 +66,161 @@ char *as_string(PyObject *object)
+    if (py_attr != NULL && PyUnicode_Check(py_attr)) {
+        tmp = as_embedded_string(py_attr, &temp_bytes);
+        if (tmp != NULL && temp_bytes != NULL) {
+            if (size <= bufferLength) {
+                size = strlen(tmp) + 1;
+                strcpy(buffer, tmp);
+            }
```
I think your first approach is sensible. Moreover it is safer if there is no terminated zero.

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -65,14 +66,161 @@ char *as_string(PyObject *object)
+
+char *string_buf_from_offset(void *buf, size_t offset)
+{
+    return (char *)(((size_t)(void *)buf) + offset);
```
Will experiment with that. I know what MS VS C++ become too temperamental with that kind of casting. Will see how CGO will fares.

iglendd replied:
Seems to be working with CGO. Thank you.

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -65,14 +66,161 @@ char *as_string(PyObject *object)
+    return (char *)(((size_t)(void *)buf) + offset);
+}
+
+size_t string_buf_from_offset_len(size_t bufLength, size_t offset)
+{
+    return bufLength > offset ? bufLength - offset : 0;
+}
+
+size_t copy_attr_as_string_at(PyObject *object, const char *attributeName, void *buf, size_t bufOffset,
+                              size_t bufLength)
+{
+    return copy_attr_as_string(object, attributeName, string_buf_from_offset(buf, bufOffset),
+                               string_buf_from_offset_len(bufLength, bufOffset));
+}
```
We could and I did have it in one function but I think I have used a ternary expression and it was too large. Your suggestion is much more sensible. Will change. Thank you.

iglendd commented:
@hush-hush thank you for great code review of rather complicated PR 🙏

iglendd commented on rtloader/common/stringutils.c:
```diff
@@ -65,14 +66,161 @@ char *as_string(PyObject *object)
+            }
+            Py_XDECREF(temp_bytes);
+        }
+    } else if (py_attr != NULL && !PyUnicode_Check(py_attr)) {
+        PyErr_Clear();
+    } else {
+        PyErr_Clear();
+    }
```
I thought that I have replied yesterday. Yes, you are absolutely correct. It looks obvious now 🤦‍♂️

Kaderinho reviewed (commented):
Thank you for this 👍 

Have a couple of nits/questions: `diangosisCounters` -> `diagnosisCounters` (unless diangosis is a word I don't know)

Also have some questions:
- The `diagnose` implementation in `pkg/diagnose/check.go` recreates the components needed to do a check (aggregator, forwarder, ...) but will in the future directly get the diagnose from the running agent service ? The current version correspond more to the local/cli diagnose of a check ?
- Is there a plan to include [HTTP tracing](https://pkg.go.dev/net/http/httptrace) ([code](https://github.com/DataDog/datadog-agent/blob/3527a4381ecbb1ce26dfd00949f2ccda2bea577d/pkg/diagnose/connectivity/hooks.go#L32)) in the `diagnose()` function later ? I think it can provide useful information and some remediation steps in some specific cases.

iglendd commented on pkg/diagnose/runner_test.go:
```diff
@@ -8,22 +8,71 @@ package diagnose
+	outSuitesDiagnoses := RunAll(diagCfg)
+	assert.Len(t, outSuitesDiagnoses, 1)
+	assert.Equal(t, outSuitesDiagnoses[0].SuiteDiagnoses, inDiagnoses)
+}
```
Yes. I did. And further improve include/exclude algorithm to have both applicable (per comment above)

iglendd replied:
> Thank you for this 👍
> 
> Have a couple of nits/questions: diangosisCounters -> diagnosisCounters (unless diangosis is a word I don't know)
> 
> Also have some questions:
> 
> The diagnose implementation in pkg/diagnose/check.go recreates the components needed to do a check (aggregator, forwarder, ...) but will in the future directly get the diagnose from the running agent service ? The current version correspond more to the local/cli diagnose of a check ?
> Is there a plan to include [HTTP tracing](https://pkg.go.dev/net/http/httptrace) ([code](https://github.com/DataDog/datadog-agent/blob/3527a4381ecbb1ce26dfd00949f2ccda2bea577d/pkg/diagnose/connectivity/hooks.go#L32)) in the diagnose() function later ? I think it can provide useful information and some remediation steps in some specific cases.
> 
> 
Yes, exactly. In the future I will get diagnose from running agent. This is the reason for --local command line option. I just implemented it first because it was easier and because I need to unblock @alopezz.

Yes for HTTP tracing. One of the reason I "removed" it from the implementation because I had a problem to suppress intermediate output. But it has lot of value indeed. I am adding this a the JIRA for backlog (see current backlog here https://datadoghq.atlassian.net/jira/software/projects/ASUP/boards/4915/backlog

iglendd commented on pkg/diagnose/README.md:
```diff
@@ -1,26 +1,64 @@
+					diagnoses = append(diagnoses, diagnosis.Diagnosis{
+						Result:      diagnosis.DiagnosisFail,
+						Name:        name,
+						Diagnosis:   fmt.Sprintf("Connection to `%s` is falied\n%s", logURL, report),
```
Thank you. Good eye 🙏

iglendd commented on pkg/epforwarder/epforwarder.go:
```diff
@@ -190,6 +198,53 @@ func (s *defaultEventPlatformForwarder) SendEventPlatformEvent(e *message.Messag
+				Result:      diagnosis.DiagnosisFail,
+				Category:    desc.category,
+				Name:        name,
+				Diagnosis:   fmt.Sprintf("Connection to `%s` is falied", url),
```
Thank you. Fixed.

iglendd commented on pkg/diagnose/connectivity/trace_request.go:
```diff
@@ -24,10 +24,65 @@ import (
+
 // RunDatadogConnectivityDiagnose sends requests to all known endpoints for all domains
 // to check if there are connectivity issues between Datadog and these endpoints
 func RunDatadogConnectivityDiagnose(writer io.Writer, noTrace bool) error {
```
That is one of my main question to you guys.  This is what I have said in https://github.com/DataDog/datadog-agent/pull/16719#discussion_r1184442552

> This is where is probably made a wrong decision from the very beginning. I have decided, afraid to change/break existing CLI, to keep "new" diagnose CLI interface separate from what we already have. I did not know exactly how "far" I will or should I go and certainly did not want to jeopardize of what is already there. I feel right now, and I have implied that in the PR comment, that I should not have "all" as segregated CLI command and instead integrate existing CLI commands into existing framework at implementation and at CLI level. In that case your suggestion would be even relevant in a sense that that would the only possible way to deal with that. May be I should take another swing into that (in a separate PR). Not sure if I explain it well though.

I can delay this PR merge and re-do CLI to eliminate "all" and other more specific sub-diagnoses but originally I was afraid to touch it because I did not know where and how I endup with the implementation.

iglendd commented on pkg/diagnose/connectivity/trace_request.go:
```diff
@@ -24,10 +24,65 @@ import (
+					d.Diagnosis = fmt.Sprintf("Connectivity to `%s` is Ok\n%s", logURL, report)
+				} else {
+					d.Result = diagnosis.DiagnosisFail
+					d.Diagnosis = fmt.Sprintf("Connection to `%s` is falied\n%s", logURL, report)
```
Fixing it, thank you

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+}
+
+// Output summary
+func outputSummary(w io.Writer, c diangosisCounters) {
```
yep, I am bad speller or rather in this case fat fingers

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+}
+
+// Overall running statistics
+type diangosisCounters struct {
```
yep, fixing it

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+	fmt.Fprint(w, "\n")
+}
+
+func incrementCounters(r diagnosis.DiagnosisResult, c *diangosisCounters) {
```
the same as below (commenting from bottom up)

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+
+	fmt.Fprintf(w, "=== Starting diagnose ===\n")
+
+	var c diangosisCounters
```
the same as below

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+
+			// Skipping not enabled diagnosis
+			if d.Result == diagnosis.DiagnosisNotEnable {
+				outputSkippedDiangosis(w, diagCfg, ds.SuitName, c, &suiteAlreadyReported, &lastDot)
```
fixed, thank you

iglendd commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +17,307 @@ import (
+	fmt.Fprint(w, "\n")
+}
+
+func outputSkippedDiangosis(w io.Writer, cfg diagnosis.DiagnoseConfig, suitName string, c diangosisCounters, suiteAlreadyReported *bool, lastDot *bool) {
```
done, thank you

iglendd commented on cmd/agent/subcommands/diagnose/command.go:
```diff
@@ -34,11 +38,27 @@ const (
+	// run as current user, value of the --run-as-user flag
+	runAsUser bool
+
+	// run diagnose on other processes, value of --remote-diagnose flag
```
Good catch. It is leftover comment. I am planning to have something in future to allow to collect diagnose from Process, Trace and System Probe processes. But it is not settled in my mind. This was leftover from that which I did not notice. Thank you again.

iglendd commented on pkg/diagnose/diagnosis/loader.go:
```diff
@@ -5,21 +5,102 @@
+}
+
+// --------------------------------
+// Register a Metadata diagnosis that will be called on diagnose
```
Good catch. Wrong comment at this moment. Addressed. Thank you.

iglendd replied:
> I think that end users ~= people with dev inclinations in our case.
imho, we should at least mention diagnose all and maybe the useful flag (--list, --include, --exclude).

Indeed. I want to agree on CLI interface first. See few comments above or below.

iglendd commented on pkg/diagnose/README.md:
```diff
@@ -1,26 +1,64 @@
-This package is used to register and run useful connectivity diagnosis on the agent.
-
-## Running all diagnosis
+This package is used to register and agent diagnostics which validates various aspects of agent installation, configuration and run-time environment.
```
Good catch. Addressed. Thank you.

iglendd commented on cmd/agent/subcommands/diagnose/README.md:
````diff
@@ -0,0 +1,50 @@
+# ```agent diagnose``` CLI command
+
+```agent diagnose``` command is one of the Agent CLI commands ([agent other commands](https://docs.datadoghq.com/agent/guide/agent-commands/?tab=agentv6v7#other-commands)). It run  internal self-diagnostic tests and outputs their results and if problems are found report them (and in some cases remediation steps). 
````
all set, thank you

iglendd committed:
Code review related changes

* Spelling mistakes
* Few small bugs
* Better coding style

iglendd committed:
Code review related changes

hush-hush commented on pkg/diagnose/runner.go:
```diff
@@ -17,8 +18,303 @@ import (
+	success        int
+	fail           int
+	warnings       int
+	unexpected_err int
```
Golang enforce camelCase:
```suggestion
	unexpectedErr int
```

iglendd replied:
Done (in the next commit). Thank you.

hush-hush commented on rtloader/common/stringutils.c:
```diff
@@ -44,10 +45,12 @@ char *as_string(PyObject *object)
-#else
-    PyObject *temp_bytes = NULL;
 
+    temp_bytes = object;
+    retval = tmp;
+    Py_INCREF(temp_bytes);
+    *stringObject = temp_bytes;
```
`temp_bytes` isn't needed here.
```suggestion
    retval = tmp;
    Py_INCREF(object);
    *stringObject = object;
```

hush-hush commented on rtloader/common/stringutils.c:
```diff
@@ -8,30 +8,31 @@
  * \return A standard C string (NULL terminated character pointer)
- *  The returned pointer is allocated from the heap and must be
- * deallocated (free()ed) by the caller
+ *  The returned pointer is embedded within PyObject. When string pointer
+ *  is not needed corresponding PyObject reference needs to be decremented
```
WDYT ?
```suggestion
 *  The returned pointer is embedded within a PyObject. When the returned string pointer
 *  is no longer needed, `stringObject` reference needs to be decremented
```

hush-hush reviewed (changes requested):
Thanks for answering all my comments. Added a few last details but I think we're good.

@truthbk Would you mind taking a quick look at `rtloader/common/stringutils.c`. Just to be on the safe side since it's a very sensitive peace of code ?

truthbk reviewed (commented):
Looking great, added some small thoughts. Thank you Len! 🙇

truthbk reviewed (commented):
Looking great, added some small thoughts. Thank you Len! 🙇

iglendd commented on rtloader/include/rtloader_types.h:
```diff
@@ -72,6 +72,40 @@ typedef struct event_s {
+// Self-contained array of diagnosis. All pointers point to portion of memory contained
+// completely within the diagnoses_t buffer. size_t or pointers are needed to avoid Go vs. CGO misalignments
+typedef struct diagnoses_s {
+    size_t byteCout;
```
of course, thank you, fixed

iglendd commented on rtloader/include/rtloader_types.h:
```diff
@@ -72,6 +72,40 @@ typedef struct event_s {
+
+// Self-contained array of diagnosis. All pointers point to portion of memory contained
+// completely within the diagnoses_t buffer. size_t or pointers are needed to avoid Go vs. CGO misalignments
+typedef struct diagnoses_s {
```
I changed it. I did not touch Go code type though. Not sure if it makes sense overall.

iglendd commented on pkg/diagnose/diagnosis/loader.go:
```diff
@@ -5,21 +5,102 @@
 		log.Warnf("Diagnosis %s already registered, overriding it", name)
 	}
-	DefaultCatalog[name] = d
+	MetadataAvailCatalog[name] = d
```
We have decided to unify other diagnose tests. When I have started I was not sure that was a right thing and tried to keep "legacy" code AS IS. But in the end I did not like it and discuss it with @hush-hush and @Kaderinho and we have decided to make one type of diagnose. I will unified CLI and catalogs as well.

iglendd committed:
Code review related changes

iglendd committed:
Fix compilation problem

vickenty commented on rtloader/common/stringutils.c:
```diff
@@ -8,30 +8,36 @@
+PyObject *yload = NULL;
+PyObject *ydump = NULL;
+PyObject *loader = NULL;
+PyObject *dumper = NULL;
```
Please don't reformat the code you did not change.

iglendd replied:
I didn't do it actually, not directly or willingly at least . Kind of. My changes to the file failed CLANG formatting tests in CI and I was forced to run CLANG formatting on the file. I have just tested by putting original formatting in place and re-running CLANG template and it did change original code to the current formatting. I am not sure why this was not done earlier or perhaps some CLANG formatting change and I theoretically can put original code back but it may suffer the same fate next time.

vickenty commented on pkg/diagnose/check.go:
```diff
@@ -0,0 +1,122 @@
+				diagnoses = append(diagnoses, diagnosis.Diagnosis{
+					Result:    diagnosis.DiagnosisUnexpectedError,
+					Name:      instanceName,
+					Category:  instanceName,
```
I was wondering what is the purpose of `Category`, if it is always the same as `Name`?

iglendd replied:
The set of Diagnosis structs (Diagnoses) could come from many places. I have added Category fields to potentially differentiate a very large Diagnosis set returned by a singe Diagnose function. In this particular case (for checks), after discussion with Integration team we have decided that they did not want to provide it (seems to be superficial for at least their current though process) hence in the code above I automatically generate it. It is different from a name, for example when a single Check's Diagnose method may provide a number of Diagnoses and they have to be differentiated by name (but their category will be the same). It is important fact that at static/compile time the meta information about a Diagnosis (name and category or description) is not set as far as registration of Diagnose method is concerned.

vickenty commented on rtloader/common/stringutils.c:
```diff
@@ -65,14 +73,131 @@ char *as_string(PyObject *object)
+    }
+
+    PyObject *py_attr = PyObject_GetAttrString(object, attributeName);
+    if (py_attr != NULL && PyUnicode_Check(py_attr)) {
```
Would it make sense to accept PyString too in Py2?

iglendd replied:
I actually do not understand nuances of the question. Can you please elaborate?

vickenty commented on rtloader/common/stringutils.c:
```diff
@@ -65,14 +73,131 @@ char *as_string(PyObject *object)
+    if (py_attr != NULL && PyUnicode_Check(py_attr)) {
+        tmp = as_embedded_string(py_attr, &temp_bytes);
+        if (tmp != NULL && temp_bytes != NULL) {
+            size = strlen(tmp) + 1;
```
Would it make sense to use Python native API to get string length?

vickenty commented on rtloader/common/diagnoseutils.c:
```diff
@@ -0,0 +1,98 @@
+    return bufferSize;
+}
+
+int serialize_diagnoses(Py_ssize_t numDiagnoses, PyObject *diagnoses_list, diagnosis_set_t *diagnoses,
```
This is very cool, but if we're serializing things anyway in a byte buffer anyway, can't we just pass JSON?

vickenty replied:
I'm not proposing to change anything in CLI, only the Python/Go interface: serialize Python objects to JSON on the Python side (or in rtloader if you must, see `as_yaml` for an example), and parse into proper Go type in python/check.go `GetDignoses`. rtloader does not need to know what Diagnosis is beyond that it's a byte blob.

iglendd replied:
Fascinating suggestion which I have not recognized in the beginning. Yes, indeed it would remove tons of complexity I have introduced around various string serialization and also make the rtloader nearly completely decoupled from either side, as you have fairly noted. I have attempted to do it. I spent half a day but was not able to figure out how to make sure that the Python Diagnosis class/object is serializable as YAML (took me a while to even find the root cause - luckily I have found PyErr_Print function).

The problem is that the YAML dumper we use cannot dump custom classes. I fail with the following stack
```
Traceback (most recent call last):
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\__init__.py", line 253, in dump
    return dump_all([data], stream, Dumper=Dumper, **kwds)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\datadog_checks\base\ddyaml.py", line 47, in safe_yaml_dump_all
    return pyyaml_dump_all(
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\__init__.py", line 241, in dump_all
    dumper.represent(data)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 27, in represent
    node = self.represent_data(data)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 48, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 199, in represent_list
    return self.represent_sequence('tag:yaml.org,2002:seq', data)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 92, in represent_sequence
    node_item = self.represent_data(item)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 58, in represent_data
    node = self.yaml_representers[None](self, data)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 231, in represent_undefined
    raise RepresenterError("cannot represent an object", data)
yaml.representer.RepresenterError: ('cannot represent an object', <datadog_checks.base.utils.diagnose.Diagnosis object at 0x0000023555390130>)
```
There is a lot of advice on how to fix it on Google and StackOverflow and I have tried many but failed to find a workable solution so far. It is also architecturally not as ideal as I thought it would be since the Diagnosis object need to "prepare" itself for YAML dumping.

I will talk with @alopezz who is working on Python side if he think it is indeed a superior approach and perhaps he can provide the proper setup for a returned Diagnosis object, than I will make necessary changes elsewhere.

vickenty replied:
There are many ways to go about it, but the simplest thing perhaps would be to have contract of Python's check GetDiagnoses to return JSON with specific keys, and handle the serialization on the Python side entirely. It then would be trivial to plug a custom serializer into JSON.dumps: https://gist.github.com/vickenty/8539570fd291e424c4beffb87cec292c#file-d-py

KSerrania reviewed (approved):
LGTM for Agent Platform-owned files.

iglendd committed:
Replaced custom serialization from Python through RTLoaded(CGO) to Go array of Diagnosis via JSON serialization

vickenty commented on pkg/collector/python/check.go:
```diff
@@ -51,6 +53,19 @@ type PythonCheck struct {
 
+// It is an identical structure to the diagnosis.Diagnosis. It is possible
+// that in future diagnosis.Diagnosis will have its own JSON mapping but
+// at this stage this single usa case is insufficient to change base struct type
```
This comment is slightly misleading, because we can't deserialize the base struct directly even if we wanted since it contains an `error` type (which JSON will refuse to deserialize).

But we can avoid having to duplicate the struct still: 1) as long as struct field names match JSON, annotations are not necessary, and 2) we can embed diagnosis into this struct and add only override the error field. The rest will get populated in the right places automatically: https://go.dev/play/p/RV6JfmKH-yo

iglendd replied:
I was not aware of these nuances (thank you for providing playground code). I will try to change implementation to be more inline with this suggestion. Thank you.

vickenty commented on rtloader/three/three.cpp:
```diff
@@ -536,6 +537,38 @@ char **Three::getCheckWarnings(RtLoaderPyObject *check)
+    ret = as_string(result);
+    if (ret == NULL) {
+        // as_string clears the error, so we can't fetch it here
+        setError("error converting 'get_diagnoses' result to string");
```
Would it make sense to return the errors from this function to the caller (to be converted to a diagnosis struct perhaps) instead of setting it on the check? IIUIC this error message shows up on the agent status page, where it may be confusing since the check itself may be functioning normally.

iglendd replied:
Good suggestion. Let me try to implement it. I was reluctant to do much reporting these kind of error transfer because I was followed existing patterns and was a bit afraid to deviate too much from them.

vickenty commented:
Would be nice to add a test for this too.

iglendd commented on pkg/collector/python/check.go:
```diff
@@ -330,6 +345,80 @@ func (c *PythonCheck) ID() check.ID {
+
+	// Copy it over (in future we may be able to deserialize
+	// directly into target data)
+	var diagnoses []diagnosis.Diagnosis
```
Good idea. I have tried diagnoses := make([]diagnosis.Diagnosis, len(diagnosesJSON)) which forced me to developed indexed driven enumeration and access which is probably even less performant. I am still learning Go :(. Thank you.

vickenty replied:
FWIW, assigning by index is a frequent pattern, and I don't think there would be any perf penalty to doing so.

For example, and there are many more throughout the code:
https://github.com/datadog/datadog-agent/blob/7b6d467eebaf7ad716d7cbc61f3a4cd43824de41/pkg/config/environment_containers.go#L144-L147

iglendd replied:
Would be interesting to see. Hopefully optimizer would eliminate indirection in memory access since I would have to do it as many times as fields number. But it also look worse code wise.

iglendd commented on rtloader/three/three.cpp:
```diff
@@ -536,6 +537,38 @@ char **Three::getCheckWarnings(RtLoaderPyObject *check)
+
+    result = PyObject_CallMethod(py_check, func_name, NULL);
+    if (result == NULL || !PyUnicode_Check(result)) {
+        setError("error invoking 'get_diagnoses' method: " + _fetchPythonError());
```
I have borrowed the code from the run method. But I also think it is wrong to report error if it is NULL since it is a valid state.

iglendd replied:
And what happens if a check's get diagnoses implementation throws exception? Is base wrapper returns internal error diagnosis?

alopezz replied:
As far as I know, `NULL` as a return value on the C-API always signals an error (even if there was no return value, that becomes `None` in Python, which in the C-API has a [special `Py_None` value](https://docs.python.org/3/c-api/none.html#c.Py_None)).

At any rate, the only valid response at this level should be a Python string, anything else being invalid. If `get_diagnoses()` returns no diagnoses that would probably still be `"[]"`. The base check will in theory guarantee that as long as individual checks don't override the `get_diagnoses()` method directly and use the framework we're building in https://github.com/DataDog/integrations-core/pull/14394. But we may not get the implementation fully right, or someone might mistakenly override the `get_diagnoses()` method and trigger such an error. Again, it's not very likely, but it doesn't hurt to handle those error cases properly.

iglendd replied:
Perfect, thank you

iglendd committed:
Code review related changes

Unit test, better deserialization

iglendd committed:
Fixed build problem

iglendd committed:
Code review related changes

- RTLoader internal failure reported in the diagnoses format
- Made unit tests TestLoadWheelCheck independent from TestConfigureDeprecated (it was hidden dependency)

iglendd committed:
Fix unit test build problem

iglendd committed:
Fixed failed unit test

iglendd committed:
Fix build problem

iglendd committed:
Fix build problem

truthbk reviewed (approved):
Simpler is better in this case. Great suggestion by @vickenty and thanks for implementing it. There is an important performance penalty but this code is not called frequently and when it is there are no strong perf requirements. Less C code is a definitely a net-positive here.

iglendd committed:
Code review related changes

iglendd commented on rtloader/rtloader/rtloader.cpp:
```diff
@@ -55,3 +59,11 @@ void RtLoader::free(void *ptr)
+char *RtLoader::_createInternalErrorDiagnoses(const char *errorMessage)
+{
+    std::string getDiagnoseFailureDiagnoses
+        = std::string(GET_DIANGOSES_FAILURE_DIAGNOSES_BEGIN) + errorMessage + GET_DIANGOSES_FAILURE_DIAGNOSES_END;
```
Good point. I have added simple escaping and log the string if deserialization fails. Regarding function signature change I am a bit ambivalent at this point. I like the simplicity of the current signature (and it also follows exiting rtloader calls patterns). I also would like to see in practice if what I have now is not sufficient and I do not see a problem to change it. Please checkout last commit.

vickenty commented on pkg/collector/python/check.go:
```diff
@@ -330,6 +338,58 @@ func (c *PythonCheck) ID() check.ID {
+	var diagnosesWrap []diagnosisJSONSerWrap
+	err = json.Unmarshal([]byte(strDiagnoses), &diagnosesWrap)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse diagnoses JSON for %s: %s. JSON: %s", c.id, err, strDiagnoses)
```
Would it make sense to use `%q` to quote any special characters that the json string may contain? This would avoid breaking log line with newlines, and also make it easier to spot any funny characters that may be causing parse errors in the first place.

```suggestion
		return nil, fmt.Errorf("failed to parse diagnoses JSON for %s: %s. JSON: %q", c.id, err, strDiagnoses)
```

vickenty reviewed (approved):
Left a couple of suggestions, but otherwise looks good.

iglendd committed:
Code review related changes

iglendd committed:
Merge remote-tracking branch 'origin/main' into len.gamburg/agent-diagnose-ext-first-cut

iglendd committed:
Merge remote-tracking branch 'origin/main' into len.gamburg/agent-diagnose-ext-first-cut

iglendd committed:
Merge remote-tracking branch 'origin/main' into len.gamburg/agent-diagnose-ext-first-cut

iglendd committed:
Fix latest main merge problems

iglendd committed:
Merge remote-tracking branch 'origin/main' into len.gamburg/agent-diagnose-ext-first-cut

iglendd committed:
Still solving merge problems

iglendd committed:
Still fixing merge issues

iglendd commented:
Ok finally rebased, all tests are passing, all reviewers passed (two team did not responded on a few requests) - merging accordingly.
